### PR TITLE
Update documentation on authoring type providers

### PR DESCRIPTION
**Pull request by dsyme**
# Revamp documentation for authoring type providers

## Summary

The document on authoring type providers was an out-of-date mess with incorrect formatting.  This gets it closer to par

## Details

* Update code to match latest TypeProvider SDK
* Clean up indentation of code (which was nearly all wildly off)
* List three different type providers

## Suggested Reviewers

@cartermp review please

### CODE REVIEW

**Comment by dsyme** [comment]
The document was a sufficient mess that we should probably just merge fairly quickly then refine (so we can see it in situ)

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md [hunk 1]
```diff
@@ -14,12 +14,17 @@ ms.assetid: 82bec076-19d4-470c-979f-6c3a14b7c70a
...
 
-The type provider mechanism in F# 3.0 is a significant part of its support for information rich programming. This tutorial explains how to create your own type providers by walking you through the development of several simple type providers to illustrate the basic concepts. For more information about the type provider mechanism in F#, see [Type Providers](index.md).
+The F# ecosystem contains a range of type providers for commonly used Internet and enterprise data services. For example:
 
-F# 3.0 contains several built-in type providers for commonly used Internet and enterprise data services. These type providers give simple and regular access to SQL relational databases and network-based OData and WSDL services. These providers also support the use of F# LINQ queries against these data sources.
+- [FSharp.Data](http://fsharp.github.io/FSharp.Data/) includes type providers for JSON, XML, CSV and HTML document formats
```
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: nit: we're using https links whenever they're accepted. I just fixed all of them in the F# directory yesterday.

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md [hunk 2]
```diff
@@ -52,7 +56,8 @@ Type providers are best suited to situations where the schema is stable at runti
 
 
 ## A Simple Type Provider
-This sample is Samples.HelloWorldTypeProvider in the `SampleProviders\Providers` directory of the [F# 3.0 Sample Pack](https://fsharp3sample.codeplex.com) on the Codeplex website. The provider makes available a "type space" that contains 100 erased types, as the following code shows by using F# signature syntax and omitting the details for all except `Type1`. For more information about erased types, see [Details About Erased Provided Types](#details-about-erased-provided-types) later in this topic.
+
+This sample is Samples.HelloWorldTypeProvider similar to the samples in the `examples` directory of the [F# Type Provider SDK](https://https://github.com/fsprojects/FSharp.TypeProviders.SDK/). The provider makes available a "type space" that contains 100 erased types, as the following code shows by using F# signature syntax and omitting the details for all except `Type1`. For more information about erased types, see [Details About Erased Provided Types](#details-about-erased-provided-types) later in this topic.
```
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: extra https in the first link.

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md [hunk 3]
```diff
@@ -1141,81 +1093,57 @@ type Service = ODataService<" https://services.odata.org/Northwind/Northwind.svc
...
 ```
   devenv /debugexe fsc.exe script.fsx
 ```
 
   You can use print-to-stdout logging.
-<br />
 
 
 ## See Also
 [Type Providers](index.md)
```
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: you'll need to add two spaces after this link so the new link shows up in a new line

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md [hunk 4]
```diff
@@ -1141,81 +1093,57 @@ type Service = ODataService<" https://services.odata.org/Northwind/Northwind.svc
...
 
 ## Rules and Limitations
 When you write type providers, keep the following rules and limitations in mind.
 
 
-- `Provided types must be reachable.`
-<br />  All provided types should be reachable from the non-nested types. The non-nested types are given in the call to the `TypeProviderForNamespaces` constructor or a call to `AddNamespace`. For example, if the provider provides a type `StaticClass.P : T`, you must ensure that T is either a non-nested type or nested under one.
-<br />  For example, some providers have a static class such as `DataTypes` that contain these `T1, T2, T3, ...` types. Otherwise, the error says that a reference to type T in assembly A was found, but the type couldn't be found in that assembly. If this error appears, verify that all your subtypes can be reached from the provider types. Note: These `T1, T2, T3...` types are referred to as the *on-the-fly* types. Remember to put them in an accessible namespace or a parent type.
-<br />
+### Provided types must be reachable.
```
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: no punctuation in headings

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md [hunk 5]
```diff
@@ -1085,52 +1048,41 @@ When you write your own type providers, you might want to use the following addi
...
-<br />
+### Exceptions and Diagnostics from Type Providers
 
+All uses of all members from provided types may throw exceptions. In all cases, if a type provider throws an exception, the host compiler attributes the error to a specific type provider.
+
+- Type provider exceptions should never result in internal compiler errors.
+
+- Type providers can't report warnings.
+
+- When a type provider is hosted in the F# compiler, an F# development environment, or F# Interactive, all exceptions from that provider are caught. The Message property is always the error text, and no stack trace appears. If you’re going to throw an exception, you can throw the following examples: `System.NotSupportedException`, `System.IO.IOException`,  `System.Exception`
```
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: ideally the apis would use xref links to go to the API docs:
`<xref:System.NotSupportedException>`, etc. 

Remove extra space after the last comma

**Review by mairaw — COMMENTED** [review]
Left a few comments.

Re. this:
> The document was a sufficient mess that we should probably just merge fairly quickly then refine (so we can see it in situ)

You'll be able to see how the content looks like on the staging site before merging. My preference would be for us to validate that staging looks good before merging.

**Comment by dsyme** [comment]
@mairaw Super, thanks!

> You'll be able to see how the content looks like on the staging site before merging. My preference would be for us to validate that staging looks good before merging.

OK, great.

**Review by cartermp — APPROVED** [review]
LGTM, pending it "looking right" on staging. @mairaw we used to have a bot which would give us the link here. Can we access it through some other means?

**Comment by mairaw** [comment]
@cartermp the build checks here on the PR include a link to the build report. Click on Details from the OpenPublishing.Build check.

Sometimes, the build will have and only have the topics you changed in there and sometimes like right now, it shows a bunch of crap. So you have to open one of those links and paste the path you want to validate. 😢 

Here's the direct link: https://review.docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider?branch=pr-en-us-4518

Feel free to merge it if you think it's ready @cartermp.

**Comment by dsyme** [comment]
I don't seem to have access to https://review.docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider?branch=pr-en-us-4518.  I tried signing in with my Microsoft account.  Should I use a different account?

**Comment by dsyme** [comment]
OK, I can access the review site, it LGTM

**Comment by cartermp** [comment]
Ditto. Thanks @dsyme!

**Pull request merged**